File: packages/sdk-elixir/lib/ubag/idempotency.ex
defmodule Ubag.Idempotency do
  @moduledoc "Generates ULID-style idempotency keys: 26 Crockford base32 characters."

  @alphabet ~c"0123456789ABCDEFGHJKMNPQRSTVWXYZ"

  @spec generate(integer()) :: binary()
  def generate(now_millis \\ System.system_time(:millisecond)) do
    timestamp = encode_base32(max(now_millis, 0), 10)

    random =
      for <<byte <- :crypto.strong_rand_bytes(16)>>, into: "" do
        <<Enum.at(@alphabet, Bitwise.band(byte, 31))>>
      end

    timestamp <> random
  end

  defp encode_base32(value, length) do
    {chars, _} =
      Enum.reduce(1..length, {[], value}, fn _, {acc, remaining} ->
        char = Enum.at(@alphabet, rem(remaining, 32))
        {[char | acc], div(remaining, 32)}
      end)

    chars |> List.to_string()
  end
end

File: packages/sdk-swift/Sources/Ubag/Idempotency.swift
import Foundation

/// Generates ULID-style idempotency keys: 26 Crockford base32 characters.
public enum IdempotencyKey {
    private static let alphabet = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ")

    public static func generate(nowMillis: Int64? = nil) -> String {
        let timestamp = nowMillis ?? Int64(Date().timeIntervalSince1970 * 1000)
        var characters = encodeBase32(max(timestamp, 0), length: 10)
        for _ in 0..<16 {
            characters.append(alphabet[Int.random(in: 0..<32)])
        }
        return String(characters)
    }

    private static func encodeBase32(_ value: Int64, length: Int) -> [Character] {
        var buffer = [Character](repeating: "0", count: length)
        var remaining = value
        var index = length - 1
        while index >= 0 {
            buffer[index] = alphabet[Int(remaining % 32)]
            remaining /= 32
            index -= 1
        }
        return buffer
    }
}
